### ISSUE TITLE
Soundness bug from exhaustivity of GADTs and polymorphic variants

### ISSUE THREAD

**Issue by lpw25** [ocaml/ocaml#9019]
The discussion in #9012 got me thinking. It seems that the current approach to refuting cases relies on the assumption that if a pattern _, Foo is typeable then so is A, Foo if A has the right type. However, that isn't the case, which leads to a fairly nasty soundness bug:

```ocaml
type _ a_or_b =
  A_or_B : [< `A of string | `B of int] a_or_b

type _ a =
  | A : [> `A of string] a
  | Not_A : _ a

type _ b =
  | B : [> `B of int ] b

let f (type x) (a : x a) (b : x b) (a_or_b : x a_or_b) (x : x) =
  match a, a_or_b, b, x with
  | Not_A, _, B, `B i -> print_int i
  | _, A_or_B, _, `A s -> print_string s

let segfault = f A B A_or_B (`B 0)
```

**Comment by lpw25** [comment]
The problem above is that the case
```ocaml
| A, A_or_B, B, `B _ ->
```
does not type-check, even though:
```ocaml
| _, A_or_B, B, `B _ ->
```
does.

**Comment by garrigue** [comment]
To clarify, this relies on the incompleteness of GADT type for polymorphic variants?

**Comment by lpw25** [comment]
Yeah, it's because you can't add a second equation for a polymorphic variant type.

**Comment by garrigue** [comment]
I see, we should at least make it compatible.

**Comment by lpw25** [comment]
Slightly simpler version:

```ocaml
type _ a_or_b =
  A_or_B : [< `A of string | `B of int] a_or_b

type _ a =
  | A : [> `A of string] a
  | Not_A : _ a

let f (type x) (a : x a) (a_or_b : x a_or_b) (x : x) =
  match a, a_or_b, x with
  | Not_A, A_or_B, `B i -> print_int i
  | _, A_or_B, `A s -> print_string s

let segfault = f A A_or_B (`B 0)
```

**Comment by garrigue** [comment]
Thanks for the example. I made an attempt at fixing that for polymorphic variants, since we know they have an incompleteness problem, but I'm not sure how general it is.

**Comment by garrigue** [comment]
This was fixed by #9154
